infra/presubmit.py

```diff
@@ -0,0 +1,312 @@
+
+
+class ProjectYamlChecker:
+  """Checks for a project.yaml file."""
+
+  # Sections in a project.yaml and the constant values that they are allowed
+  # to have.
+  SECTIONS_AND_CONSTANTS = {
+      'sanitizers': {'address', 'none', 'memory', 'undefined', 'dataflow'},
+      'architectures': {'i386', 'x86_64'},
+      'engines': {'afl', 'libfuzzer', 'honggfuzz', 'dataflow'}
+  }
+
+  # Note: this list must be updated when we allow new sections.
+  VALID_SECTION_NAMES = [
+      'homepage', 'primary_contact', 'auto_ccs', 'sanitizers', 'architectures',
+      'disabled', 'view_restrictions', 'coverage_extra_args', 'vendor_ccs'
+  ]
+
+  # Note that some projects like boost only have auto-ccs. However, forgetting
+  # primary contact is probably a mistake.
+  REQUIRED_SECTIONS = ['primary_contact']
+
+  def __init__(self, filename):
+    self.filename = filename
+    with open(filename) as file_handle:
+      self.data = yaml.safe_load(file_handle)
+
+    self.success = True
+
+  def do_checks(self):
+    """Do all project.yaml checks. Return True if they pass."""
+    if self.is_disabled():
+      return True
+
+    checks = [
+        self.check_project_yaml_constants, self.check_required_sections,
+        self.check_valid_section_names, self.check_valid_emails
+    ]
+    for check_function in checks:
+      check_function()
+    return self.success
+
+  def is_disabled(self):
+    """Is this project disabled."""
+    return self.data.get('disabled', False)
+
+  def print_error(self, message, *args):
+    """Print an error message and set self.success to False."""
+    self.success = False
+    message = message % args
+    print('Error in %s: %s' % (self.filename, message))
+
+  def check_project_yaml_constants(self):
+    """Check that certain sections only have certain constant values."""
+    for section, allowed_constants in self.SECTIONS_AND_CONSTANTS.items():
+      if section not in self.data:
+        continue
+      actual_constants = self.data[section]
+      for constant in actual_constants:
+        if constant not in allowed_constants:
+          self.print_error('%s (in %s section) is not one of %s', constant,
+                           section, allowed_constants)
+
+  def check_valid_section_names(self):
+    """Check that only valid sections are included."""
+    for name in self.data:
+      if name not in self.VALID_SECTION_NAMES:
+        self.print_error('%s not a valid section name (%s)', name,
+                         self.VALID_SECTION_NAMES)
+
+  def check_required_sections(self):
+    """Check that all required sections are present."""
+    for section in self.REQUIRED_SECTIONS:
+      if section not in self.data:
+        self.print_error('No %s section.', section)
+
+  def check_valid_emails(self):
+    """Check that emails are valid looking."""
+    # Get email addresses.
+    email_addresses = []
+    for section in ['auto_ccs', 'primary_contact']:
+      email_addresses.extend(self.data.get(section, []))
```

Comment:
This doesn't do teh right thing -- `self.data.get("primary_contact")` is a string, not a list, so `extend()` treats it as a sequence of chars

Reply:
Fixing!

Reply:
This should be now fixed with https://github.com/google/oss-fuzz/commit/f06afc1ebe31ab45c4379a7b6f9726c4f367d894, please sync to head.